### ISSUE TITLE
Apparently recoverable InternalError on ARAX "kitchen sink" query

### ISSUE THREAD

**Issue by saramsey** [RTXteam/RTX#1734]
So, when running this DSL query on `arax-backup.rtx.ai` (production devarea) last night:
```
add_qnode(name=arthritis, key=n00)
add_qnode(categories=biolink:Protein, is_set=true, key=n01)
add_qnode(categories=biolink:ChemicalEntity, key=n02)
add_qedge(subject=n00, object=n01, key=e00)
add_qedge(subject=n01, object=n02, key=e01, predicates=biolink:physically_interacts_with)
expand(edge_key=[e00,e01], kp=RTX-KG2)
overlay(action=overlay_clinical_info, observed_expected_ratio=true, virtual_relation_label=C1, subject_qnode_key=n00, object_qnode_key=n02)
filter_kg(action=remove_edges_by_continuous_attribute, edge_attribute=probably_treats, direction=below, threshold=.8, remove_connected_nodes=t, qnode_keys=[n02])
overlay(action=compute_jaccard, start_node_key=n00, intermediate_node_key=n01, end_node_key=n02, virtual_relation_label=J1)
overlay(action=predict_drug_treats_disease, subject_qnode_key=n02, object_qnode_key=n00, virtual_relation_label=P1)
resultify(ignore_edge_direction=true)
filter_results(action=limit_number_of_results, max_results=15)
return(message=true, store=false)
```
I saw an interesting entry in the logfile from the `query_controller` child process (attached), which seemed to occur in the ranker (?):
```
2021-11-16T06:56:23.056243 WARNING: [] QueryGraph appears to be circular or has a strange geometry. This might cause trouble
2021-11-16T06:56:23.056301 ERROR: [InteralError_F260] Reached loop max: 21
```
Is this expected?  It seemed to recover from the error and continue on its way. Though there were no results in the end, in this case.

From the interpreted DSL, the query graph was:
```
2021-11-16T06:56:23.058652 DEBUG: [] Query graph is {'edges': {'C1': {'constraints': [],
                  'exclude': None,
                  'object': 'n02',
                  'option_group_id': None,
                  'predicates': 'biolink:has_real_world_evidence_of_association_with',
                  'subject': 'n00'},
           'J1': {'constraints': [],
                  'exclude': None,
                  'object': 'n02',
                  'option_group_id': None,
                  'predicates': ['biolink:has_jaccard_index_with'],
                  'subject': 'n00'},
           'P1': {'constraints': [],
                  'exclude': None,
                  'object': 'n00',
                  'option_group_id': None,
                  'predicates': 'biolink:probably_treats',
                  'subject': 'n02'},
           'e00': {'constraints': [],
                   'exclude': False,
                   'object': 'n01',
                   'option_group_id': None,
                   'predicates': None,
                   'subject': 'n00'},
           'e01': {'constraints': [],
                   'exclude': False,
                   'object': 'n02',
                   'option_group_id': None,
                   'predicates': ['biolink:physically_interacts_with'],
                   'subject': 'n01'}},
 'nodes': {'n00': {'categories': None,
                   'constraints': [],
                   'ids': ['MONDO:0005578'],
                   'is_set': False,
                   'name': None,
                   'option_group_id': None},
           'n01': {'categories': ['biolink:Protein'],
                   'constraints': [],
                   'ids': None,
                   'is_set': True,
                   'name': None,
                   'option_group_id': None},
           'n02': {'categories': ['biolink:ChemicalEntity'],
                   'constraints': [],
                   'ids': None,
                   'is_set': False,
                   'name': None,
                   'option_group_id': None}}}
```

[arax-query-controller-child-process-gc92r_dt.log.gz](https://github.com/RTXteam/RTX/files/7548312/arax-query-controller-child-process-gc92r_dt.log.gz)

**Comment by saramsey** [comment]
Tagging @finnagin and @dkoslicki for their take. And since the logfile entry referenced a weird graph geometry, tagging @amykglen as well.

**Comment by edeutsch** [comment]
This is a problem in the query_graph_info module. But I don't know what the cause is. It looks like a simple query-graph

**Comment by edeutsch** [comment]
If query_graph_info is invoked after all the overlay stuff, then it may be better confused by virtual edge C1.
We could add biolink:has_real_world_evidence_of_association_with as an edge to be ignored at line 130.
BUT, I'm vaguely thinking that biolink:has_real_world_evidence_of_association_with may sometimes be a *real* edge, too?
We may want to devise a more reliable way to annotate virtual edges so that ARAX can ignore them better. Both in query_graph_info and query_graph_interpreter.

**Comment by saramsey** [comment]
FWIW, KG2.7.4 has 1,027 edges with the `biolink:has_real_world_evidence_of_association_with` predicate:

<img width="1091" alt="Screen Shot 2021-11-16 at 12 16 11 PM" src="https://user-images.githubusercontent.com/5562107/142059029-4b74fc81-e9c2-4424-b1c0-a28f86d85c70.png">

**Comment by edeutsch** [comment]
ah, if the same predicate can be used for a virtual edge and a real edge, it will cause some problems here, and it is all the more important to devise a more reliable way to annotate virtual edges

**Comment by finnagin** [comment]
So does query_graph_info always get called after DSL is run? I thought that was something that was called first before the DSL?

**Comment by edeutsch** [comment]
I was kinda wondering the same thing. I think I originally wrote it to inspect the incoming query_graph. But then somehow I later inferred that it was being run by resultify. Perhaps to decide what the essence of a result it. In short, I don't know, but it seems so.

**Comment by dkoslicki** [comment]
FWIW, since CHP/ICEES/COHD all want to use the `biolink:has_real_world_evidence_of_association_with` predicate, and this is used by our `overlay_clinical_info`, such instances of predicates for real and virtual edges will arise. So question is:
1. Are the `biolink:has_real_world_evidence_of_association_with` edges supposed to be in KG2/C
2. If so, the options are a) change the overlay name or b) change the `query_graph_info` code to ignore virtual edges. I would vote the later b). Since virtual edges are "second class citizens"

**Comment by finnagin** [comment]
Looks like this attribute that Chunyu added could be used to identify virtual edges: https://github.com/RTXteam/RTX/issues/1566#issuecomment-903036486

**Comment by finnagin** [comment]
From meeting: plan is to work on this at the next mini hackathon on  1/19

**Comment by saramsey** [comment]
Restated this query graph (as best I understood it) in modern TRAPI:
```
{"edges": {"e00": {"object": "n01",
                   "subject": "n00"},
           "e01": {"object": "n02",
                   "predicates": ["biolink:physically_interacts_with"],
                   "subject": "n01"}},
 "nodes": {"n00": {"ids": ["MONDO:0005578"],
                   "is_set": false},
           "n01": {"categories": ["biolink:Protein"],
                   "is_set": true},
           "n02": {"categories": ["biolink:ChemicalEntity"],
                   "is_set": false}}}

```

**Comment by saramsey** [comment]
The DSL also works, with slight modification:
```
add_qnode(name=arthritis, key=n00)
add_qnode(categories=biolink:Protein, is_set=true, key=n01)
add_qnode(categories=biolink:ChemicalEntity, key=n02)
add_qedge(subject=n00, object=n01, key=e00)
add_qedge(subject=n01, object=n02, key=e01, predicates=biolink:physically_interacts_with)
expand(edge_key=[e00,e01], kp=infores:rtx-kg2)
overlay(action=overlay_clinical_info, observed_expected_ratio=true, virtual_relation_label=C1, subject_qnode_key=n00, object_qnode_key=n02)
filter_kg(action=remove_edges_by_continuous_attribute, edge_attribute=probably_treats, direction=below, threshold=.8, remove_connected_nodes=t, qnode_keys=[n02])
overlay(action=compute_jaccard, start_node_key=n00, intermediate_node_key=n01, end_node_key=n02, virtual_relation_label=J1)
overlay(action=predict_drug_treats_disease, subject_qnode_key=n02, object_qnode_key=n00, virtual_relation_label=P1)
resultify(ignore_edge_direction=true)
filter_results(action=limit_number_of_results, max_results=15)
return(message=true, store=false)

```

**Comment by dkoslicki** [comment]
The `infer_utilities.py` code seems to not like the fact that no q-edge was given for the `drug_treatment_graph_expansion` part of the DSL.
```
 2025-08-07T15:18:58.422853 ERROR: [UnknownError]    qedge_id is None but QG is not empty30.219 s
   2025-08-07T15:18:58.427538 ERROR: [UncaughtARAXiError]    An uncaught error occurred: qedge_id is None but QG is not empty: ['Traceback (most recent call last):\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_query.py", line 809, in execute_processing_plan\n infer.apply(response, action[\'parameters\'])\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_infer.py", line 366, in apply\n getattr(self, \'_\' + self.__class__.__name__ + \'__\' + parameters[\'action\'])() # thank you https://stackoverflow.com/questions/11649848/call-methods-by-string\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_infer.py", line 598, in __drug_treatment_graph_expansion\n self.response, self.kedge_global_iter, self.qedge_global_iter, self.qnode_global_iter, self.option_global_iter = iu.genrete_treat_subgraphs(self.response, top_scores, top_paths, qedge_id, self.kedge_global_iter, self.qedge_global_iter, self.qnode_global_iter, self.option_global_iter)\n', ' File "/mnt/data/orangeboard/production/RTX/code/ARAX/ARAXQuery/Infer/scripts/infer_utilities.py", line 117, in genrete_treat_subgraphs\n raise Exception("qedge_id is None but QG is not empty")\n', 'Exception: qedge_id is None but QG is not empty\n']
```
Changing this to specify a `qedge_id` and adding the qedge to the QG, we get the following DSL:
```
add_qnode(name=arthritis, key=n00)
add_qnode(categories=biolink:Protein, is_set=true, key=n01)
add_qnode(categories=biolink:ChemicalEntity, key=n02)
add_qedge(subject=n00, object=n01, key=e00)
add_qedge(subject=n01, object=n02, key=e01, predicates=biolink:physically_interacts_with)
add_qedge(subject=n00, object=n02, key=e02)
expand(edge_key=[e00,e01], kp=infores:rtx-kg2)
overlay(action=overlay_clinical_info, observed_expected_ratio=true, virtual_relation_label=C1, subject_qnode_key=n00, object_qnode_key=n02)
filter_kg(action=remove_edges_by_continuous_attribute, edge_attribute=probably_treats, direction=below, threshold=.8, remove_connected_nodes=t, qnode_keys=[n02])
overlay(action=compute_jaccard, start_node_key=n00, intermediate_node_key=n01, end_node_key=n02, virtual_relation_label=J1)
infer(action=drug_treatment_graph_expansion, disease_curie=MONDO:0005578, qedge_id=e02)
resultify(ignore_edge_direction=true)
```
Which unfortunately still does not work, but now for a different reason:

<img width="1345" height="650" alt="Image" src="https://github.com/user-attachments/assets/1441cf9b-231c-4aee-bf61-a50c3178e5e0" />

@chunyuma do you know why the `drug_treatment_graph_expansion` is adding qedges, and how we should properly call it from here? I _think_ the intention was to have `arthritis->Protein->ChemicalEntity`, and then a `arthritis <-[creative treats]- ChemicalEntity` kind of query.

But to be honest, I'm a touch mystified about this query. For example, why is it trying to filter on `probably_treats` _before_ doing the creative treats part. Maybe this is a hold-over when xDTD was triggered on all `disease->drug` like queries?

FWIW, switching the direction of edge e02 makes the query "complete", but with the following error and zero results:
```
 2025-08-07T15:36:44.062436 ERROR: [UncaughtARAXiError]    An uncaught error occurred: 'n00': ['Traceback (most recent call last):\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_query.py", line 809, in execute_processing_plan\n infer.apply(response, action[\'parameters\'])\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_infer.py", line 366, in apply\n getattr(self, \'_\' + self.__class__.__name__ + \'__\' + parameters[\'action\'])() # thank you https://stackoverflow.com/questions/11649848/call-methods-by-string\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_infer.py", line 598, in __drug_treatment_graph_expansion\n self.response, self.kedge_global_iter, self.qedge_global_iter, self.qnode_global_iter, self.option_global_iter = iu.genrete_treat_subgraphs(self.response, top_scores, top_paths, qedge_id, self.kedge_global_iter, self.qedge_global_iter, self.qnode_global_iter, self.option_global_iter)\n', ' File "/mnt/data/orangeboard/production/RTX/code/ARAX/ARAXQuery/Infer/scripts/infer_utilities.py", line 410, in genrete_treat_subgraphs\n self.resultify_and_sort(essence_scores)\n', ' File "/mnt/data/orangeboard/production/RTX/code/ARAX/ARAXQuery/Infer/scripts/infer_utilities.py", line 82, in resultify_and_sort\n self.response = resultifier.apply(self.response, resultify_params)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_resultify.py", line 122, in apply\n self._resultify(mode=mode, describe=False)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_resultify.py", line 252, in _resultify\n results = _get_results_for_kg_by_qg(kg,\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_resultify.py", line 676, in _get_results_for_kg_by_qg\n result_graphs_for_option_group = _create_result_graphs(option_group_qg, kg_node_keys_by_qg_key_collapsed,\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_resultify.py", line 1143, in _create_result_graphs\n kg_node_adj_map_by_qg_key = _get_kg_node_adj_map_by_qg_key(kg_node_keys_by_qg_key, edge_keys_by_node_pair, qg, log)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/openapi_server/../../../../ARAX/ARAXQuery/ARAX_resultify.py", line 968, in _get_kg_node_adj_map_by_qg_key\n kg_node_to_node_map[qnode_key_1][node_a][qnode_key_2].add(node_b)\n', "KeyError: 'n00'\n"]
```
basically, `expand` not liking how `infer_utilities.py` called it.

**Comment by chunyuma** [comment]
Hi @saramsey @dkoslicki,

Sorry for the the delay  in investigating the issues in the above query. 

After careful investigation, this query has the following errors or some potential problems that we need to discuss:

__1. Use a wrong `edge_attribute` in the filter_kg module__

In this query, it used `edge_attribute=probably_treats`. First, the query didn't call an action `predict_drug_treats_disease` in the overlay module before filter_kg module but it attempted filter the edges with attribute associated with "probably_treats". Second, the "probably_treats" is the old attribute name. It is now called "probability_treats".

__2. Both creative mode and normal mode are queried together__

When we design the `infer` mode, we didn't expect both `expand` and `infer` are called together. 

For instance, in most use cases of the creative mode, we use the queries like:
```
add_qnode(name=acetaminophen, key=n0)
add_qnode(categories=biolink:Gene, key=n1)
add_qedge(subject=n0, object=n1, key=e0)
infer(action=chemical_gene_regulation_graph_expansion, subject_qnode_id=n0, qedge_id=e0, regulation_type=increase, n_result_curies=1, n_paths=1)
overlay(action=compute_ngd, virtual_relation_label=N1, subject_qnode_key=n0, object_qnode_key=n1)
resultify()
filter_results(action=limit_number_of_results, max_results=30)
```
Here, the `infer` module replaces the `expand` module in the creative mode. 

If the `expand` module is used, we don't expect the `infer` module is called in the same query.

The reason why the `expand` and `infer` cannot exist together is because the expand module might generate some edges that are the same as the inferred edges that xDTD or xCRG generates. The `expand`-based edge and `infer`-based edges are the same but have different query ids, which is not allowed currently in the system.

In addition, when the `disease_curie` is used in the `infer` module, we expect this curie is the object of the edge (see [the document](https://github.com/RTXteam/RTX/blob/master/code/ARAX/Documentation/DSL_Documentation.md#inferactiondrug_treatment_graph_expansion)) instead of the subject of the edge. However, in this case, the disease `arthritis` is the subject of the query edge `e02`.

**Comment by dkoslicki** [comment]
Ah ha! Thanks for the detailed explanation @chunyuma . So the long and short of it is: don't expect to be able to use `expand` and `infer` at the same time, correct?

FWIW, Translator does want the ability in the future to mix creative queries with lookups (eg. `X-[lookup]->Y-[pathfinder]->Z-[lookup]-A`). So this issue will probably arise in the future, but likely should be its own issue instead of this one. 

For this issue, I guess the right way to go is ensure none of our tests (or things in `test_ARAX_api.py`) have both an `infer` and an `expand` in them.

**Comment by chunyuma** [comment]
That's exactly right. Another thing is that `disease_curie` parameter in the `infer` module should be the object of a query edge (see [this](https://github.com/RTXteam/RTX/blob/master/code/ARAX/Documentation/DSL_Documentation.md#inferactiondrug_treatment_graph_expansion) document). I am not sure if we need to make it more flexible.
